Orange/classification/rules.py

```diff
@@ -0,0 +1,947 @@
+    ----------
+    a : array_like
+        The source array.
+    random_seed : int
+        RNG seed.
+
+    Returns
+    -------
+    index : int
+        Index of the maximum value.
+
+    Raises
+    ------
+    ValueError : shape mismatch
+        If 'a' has got more than 2 dimensions.
+
+    Notes
+    -----
+    2-D arrays are also supported to avoid multiple RNG initialisation.
+    An array of indices corresponding to the maximum value of each row
+    is then returned.
+    """
+    if a.ndim > 2:
+        raise ValueError("argmaxrnd only accepts arrays of up to 2 dim")
+
+    random = (np.random if random_seed is None
+              else np.random.RandomState(random_seed))
+
+    def f(x): return random.choice((x == bn.nanmax(x)).nonzero()[0])
+
+    return f(a) if a.ndim == 1 else np.apply_along_axis(f, axis=1, arr=a)
+
+
+def entropy(x):
+    x = x[x != 0]
+    x = x / np.sum(x, axis=0)
+    x = -x * np.log2(x)
+    return np.sum(x)
+
+
+def likelihood_ratio_statistic(x, y):
+    x[x == 0] = 1e-5
+    y[y == 0] = 1e-5
+    lrs = np.sum(x * np.log(x / y))
+    lrs = 2 * (lrs - np.sum(x) * np.log(np.sum(x) / np.sum(y)))
+    return lrs
+
+
+def get_dist(Y, domain):
+    """
+    Determine the class distribution for a given array. Gather the
+    number of classes from the data domain.
+
+    Parameters
+    ----------
+    Y : array_like
+        The source array (observed classification).
+    domain : Orange.data.domain.Domain
+        Data domain.
+
+    Returns
+    -------
+    dist : ndarray, int
+        Distribution of classes.
+    """
+    return np.bincount(Y.astype(dtype=np.int),
+                       minlength=len(domain.class_var.values))
+
+
+def hash_dist(a):
+    """
+    For a given class distribution, calculate a hash value that can be
+    used to seed the RNG.
+
+    Parameters
+    ----------
+    a : array_like
+        The source array.
+
+    Returns
+    -------
+    hash : int
+        Hash function result.
+    """
+    return int(sha1(bytes(a)).hexdigest(), base=16) & 0xffffffff
+
+
+def rule_length(rule):
+    return len(rule.selectors)
+
+
+class Evaluator:
+    def evaluate_rule(self, rule):
+        """
+        Characterise a search heuristic.
+
+        If lower values indicate better results, return negatives to
+        correctly integrate the sorting procedure.
+
+        Parameters
+        ----------
+        rule : Rule
+            Evaluate this rule.
+
+        Returns
+        -------
+        res : float
+            Evaluation function result.
+        """
+        raise NotImplementedError("descendants of Evaluator must "
+                                  "overload method 'evaluate_rule'")
+
+
+class EntropyEvaluator(Evaluator):
+    def evaluate_rule(self, rule):
+        x = rule.curr_class_dist.astype(dtype=np.float)
+        if rule.target_class is not None:
+            x = np.array([x[rule.target_class],
+                          np.sum(x) - x[rule.target_class]])
```

Comment:
If this is called a lot, `x[list_of_indices]` makes a copy, so you can cache it in a variable to avoid making it twice.

`x.sum()`